numba/core/untyped_passes.py

```diff
@@ -1680,6 +1681,103 @@ def run_pass(self, state):
+
+        if call_fn.value not in (iter, zip):
+            return False
+
+        for arg in call_expr.args:
+            # only one arg?
+            if self.is_arg_constsized(state, arg):
+                # store the call expr to be removed later
+                self.dead_vars.append(func_ir.get_assignee(call_expr))
+                return True
+        return False
+
+    def find_dead_loops(self, state):
+        dead_loops = list()
+        cfg = compute_cfg_from_blocks(state.func_ir.blocks)
+        for loop in cfg.loops().values():
+            header = state.func_ir.blocks[loop.header]
+            for expr in header.find_exprs('iternext'):
+                expr = guard(get_definition, state.func_ir, expr.value.name)
+                if expr:
+                    if isinstance(expr, ir.Expr) and expr.op == 'getiter':
+                        # expr := getiter(value=Var)
+                        if self.is_arg_constsized(state, expr.value) or \
+                           self.match_iter(state, expr):
+                            dead_loops.append(loop)
+        return dead_loops
+
+    def remove_dead_loops(self, state, dead_loops):
+        for loop in dead_loops:
+            # change branch into a direct jump to exit
+            header = state.func_ir.blocks[loop.header]
+            term = header.terminator
+            header.insert_before_terminator(ir.Jump(term.falsebr, term.loc))
+            header.remove(term)
+
+            # iterate over header and remove "pair_first", since
+            # dead_code_elimination won't remove it
+            for inst in header.find_insts(ir.Assign):
+                expr = inst.value
+                if isinstance(expr, ir.Expr) and expr.op == 'pair_first':
+                    header.remove(inst)
+
+        dead_branch_prune(state.func_ir, state.args)
+
+        for var in self.dead_vars:
+            if var is None:
+                continue
+            for block in state.func_ir.blocks.values():
+                assign = block.find_variable_assignment(var.name)
+                if assign:
+                    block.remove(assign)
+
+    def run_pass(self, state):
+        # run as subpipeline
+        from numba.core.compiler_machinery import PassManager
+        from numba.core.typed_passes import PartialTypeInference
+        pm = PassManager("dead_loop_elimination_subpipeline")
+        if not state.flags.no_rewrites:
+            pm.add_pass(RewriteSemanticConstants, "Semantic const")
+            pm.add_pass(DeadBranchPrune, "Dead branch prune")
+            pm.add_pass(GenericRewrites, "Generic Rewrites")
+        pm.add_pass(PartialTypeInference, "performs partial type inference")
+        pm.finalize()
+        pm.run(state)
+
+        dead_loops = self.find_dead_loops(state)
+        if dead_loops:
+            self.remove_dead_loops(state, dead_loops)
+            state.func_ir.blocks = simplify_CFG(state.func_ir.blocks)
+            dead_code_elimination(state.func_ir)
```

Comment:
@stuartarchibald From first principles, I don't think that DCE should absolutely require types.  However, I looked at the code and our DCE does use typemap because it can improve alias analysis and do a better job at DCE.  Typemap is passed around to various other functions and I spot checked some of them and the story is the same there that typemap can improve accuracy but isn't 100% required.  I didn't check absolutely every call path so can't say that right now the typemap is optional but if not then it should be possible to make that happen.

Reply:
I remember having issues with DCE without typemap but don't remember the details. Ideally, it should be more conservative but work without typemap.